### DISCUSSION TITLE
Geany as alternative editor

### DISCUSSION THREAD

**Discussion by Cicorione** [speedata/publisher#610] [Q&A]
Hi @pgundlach,

I am trying to let the brand-new **[Geany LSP plugin](https://github.com/techee/geany-lsp/)** working with `schema`, and for now with the java `lemminx` server on `aarch64`. We aren't there yet, one of the reasons is because I think I am not assign properly the schema to the XML file; I believe I found two strategies to be most effective even though everything is still a mess.

**You are the expert in XML and I hope you can lead us toward the right direction. 🙏**
*Discussion here: https://github.com/techee/geany-lsp/discussions/41#discussioncomment-10479170*

#### Method 1
```
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="/home/freezr/git/publisher/schema/layoutschema-en.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="/home/freezr/git/publisher/schema/layoutschema-en.xsd" type="application/xml" schematypens="http://http://www.w3.org/2001/XMLSchema"?>
```

#### Method 2
```
<Layout xmlns="urn:speedata.de:2009/publisher/en"
	xmlns:sd="urn:speedata:2009/publisher/functions/en"
	xmlns:xi="http://www.w3.org/2001/XInclude"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="/home/cicorione/git/publisher/schema/layoutschema-en.xsd /home/cicorione/git/publisher/schema/layoutschema-en.xsd"/>
``` 

This is on of the method recommended by the [documentation](https://github.com/eclipse/lemminx/blob/main/docs/Features.md) but I get this error:

![geany-lsp](https://github.com/user-attachments/assets/5498c50d-7e29-4f2d-acb6-c59752493595)

The path looks correct to me... 

======================

The reason is that VSCode doesn't work very well on low-end hardware, aarch64 and BSD because being an electron app, on the contrary Geany is full featured, light & blazing fast and it would be another amazing alternative for working with Publisher.

**Comment by pgundlach** [comment]
I am always confused by the editor dependent solutions for locating the XSD or RNG files. I believe your line 11 in the screenshot above (using the `xsi:schemaLocation`) looks good.

XML catalogs are my favourite way of specifying the location, independent of the file. I know that the RedHat XML mode (using the lemminx server) is capable of using catalogs, so I never use the file-local setup.

(You know the catalog file, but for future reference, here it is: https://github.com/speedata/publisher/blob/develop/schema/catalog-schema-en.xml)

**Reply by Cicorione** [reply]
Thank you @pgundlach!

I am passing the `catalog` through a json file to the server:

```json
{
  "settings": {
    "xml": {
      "trace": {
        "server": "verbose"
      },
      "catalogs": ["/home/cicorione/git/publisher/schema/catalog-schema-en.xml"],
      "logs": {
        "client": true,
        "file": "/home/freezr/lsp4xml.log"
      },
      "format": {
        "splitAttributes": true,
        "joinCDATALines": false,
        "joinContentLines": false,
        "joinCommentLines": false,
        "spaceBeforeEmptyCloseTag": false,
        "enabled": true
      }
    }
  }
}
```

And from the editor I get:
```
07:45:10: Info: Aug 30, 2024 07:45:10 org.eclipse.lemminx.extensions.contentmodel.uriresolver.XMLCatalogResolverExtension setCatalogs()
Message: Adding XML catalog '/home/cicorione/git/publisher/schema/catalog-schema-en.xml' with expand system id 'file:///home/cicorione/git/publisher/schema/catalog-schema-en.xml' and root URI 'file:///home/cicorione/sp/agl/'.
```

Now I noticed the the message was cut, maybe I need to have the file `layoutschema-en.rng` on the `root URI`... 🤔

**Comment by pgundlach** [comment]
Is there a simple way to try that myself? I have installed Geany, but cannot find a way to install this LSP. Is there a tutorial on how to install the plugins?

**Reply by Cicorione** [reply]
Hi Patrick,

I made it working but I think I have to enable a couple of missing features, more info here: https://github.com/techee/geany-lsp/discussions/41#discussioncomment-10500781

The LSP plugins for Geany works only with the latest version 2.1 from Github, I have collected some notes necessary to compile it:

```
git clone https://github.com/geany/geany.git
git clone https://github.com/geany/geany-plugins.git
git clone https://github.com/techee/geany-lsp.git
autotools-dev
libgtk-3-dev
python3-docutils
intltool
liblua5.1-0-dev
libwebkit2gtk-4.1-dev
libenchant-2-dev
libgit2-dev
libvte-2.91-dev
libctpl-dev
libxml2-dev
libgtkspell3-3-dev
```

Also note that I am using the Lemminx java version because I am doing these tests on `aarch64`: 
https://download.eclipse.org/lemminx/releases/0.28.0/org.eclipse.lemminx-uber.jar

But for x86 it does exist a binary file that doesn't require java to run: 
https://download.jboss.org/jbosstools/vscode/stable/lemminx-binary/0.19.1-541/

Thanks! 🙏

**Reply by pgundlach** [reply]
I have 2.0, so I will either wait or try to find a way to get 2.1 working.

**Comment by Cicorione** [comment]
Hi @pgundlach 

We are going better:

1. works with `lemminx "java"` & `lemminx-linux`
2. Schema assign works with RNG and XSD definition
3. Autocompletion and suggestions work with some quirks (cursor ends outside the quotation marks)
4. Taskbar validation message: `issue=0` stands for "XML is valid"; he wants make the LSP plugin agnostic as much as possible.

Check this out:
![geany-lsp-02](https://github.com/user-attachments/assets/c1d29a0c-3ae3-4e15-85aa-1b6a91fe7046)

This is pretty cool cause VSCode was getting frozen on my aarch64 board, as you already noted I work across several machines, this SBC is one of them and perhaps the one where I am connected the most... 😅

Geany with LSP will give me an editor with a human-friendly GUI like VSCode, truly multiplatform and extremely light and fast. I look forward to test it extensively on FreeBSD as well!

**Comment by Cicorione** [comment]
Further changes where made, now it works like charm, and now I can work even on my aarch64 SBC, it is amazing!

Here a recap about the settings:

#### lsp.conf
```
[XML]
cmd=java -jar /your/path/org.eclipse.lemminx-uber.jar # or lemminx-linux if amd64 
initialization_options_file=/your/path/xml_init_conf.json # you decide the name
autocomplete_use_label=true
autocomplete_use_snippets=true
diagnostics_statusbar_severity=4
```

#### xml_init_conf.json
```json
{
    "settings": {
        "xml": {
            "trace": {
                "server": "verbose"
            },
            "catalogs": [
                "/your/path/schema/catalog-schema-en.xml"
            ],
            "logs": {
                "client": true,
                "file": "/your/path/lsp4xml.log"
            },
            "format": {
                "splitAttributes": true,
                "joinCDATALines": false,
                "joinContentLines": false,
                "joinCommentLines": false,
                "spaceBeforeEmptyCloseTag": true,
                "enabled": true
            }
        }
    }
}
```
